transport/http2_client.go

```diff
@@ -554,17 +569,32 @@ func (t *http2Client) handleSettings(f *http2.SettingsFrame) {
 	}
 	f.ForeachSetting(func(s http2.Setting) error {
 		if v, ok := f.Value(s.ID); ok {
-			t.mu.Lock()
-			defer t.mu.Unlock()
 			switch s.ID {
 			case http2.SettingMaxConcurrentStreams:
-				t.maxStreams = v
+				// TODO(zhaoq): This is a hack to avoid significant refactoring of the
+				// code to deal with the unrealistic int32 overflow. Probably will try
+				// to find a better way to handle this later.
+				if v > math.MaxInt32 {
+					v = math.MaxInt32
+				}
+				t.mu.Lock()
+				reset := t.streamsQuota != nil
+				ms := t.maxStreams
+				t.maxStreams = int(v)
+				t.mu.Unlock()
+				if !reset {
+					t.streamsQuota = newQuotaPool(int(v))
+				} else {
+					t.streamsQuota.reset(int(v) - ms)
```

Comment:
This still doesn't work, as far as I can tell.  The reset is atomic, but it doesn't take into account the number of streams in-flight.

If you want to swap out t.streamsQuota, you need to ensure that there are no streams in-flight.  Otherwise, you'll need to implement this in terms of acquire and/or add.

One alternative to consider: if you store the original quota size, then you can compute the delta between the new and old settings.  If new > old, you can just add it; if new < old, you can acquire the difference and just not add it back.

But that makes the handleSettings function take an arbitrarily long time while it waits for the existing streams to settle...